## Упражнения по библиотеке Numpy

In [33]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [34]:
a = np.arange(0, 10)
a[(3 < a) & (a < 8)] *= -1
print(a)

[ 0  1  2  3 -4 -5 -6 -7  8  9]


**2.** Заменить максимальный элемент случайного массива на 0

In [35]:
a = np.arange(0, 10)
a[(np.max(a))] = 0
print(a)

[0 1 2 3 4 5 6 7 8 0]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [36]:
import numpy as np

def product(a, b):
    a, b = np.meshgrid(a, b)
    result = np.column_stack((a.ravel(), b.ravel()))
    return result

a = np.array([1, 2, 3])
b = np.array(['a', 'b'])

result_product = product(a, b)

print(result_product)

[['1' 'a']
 ['2' 'a']
 ['3' 'a']
 ['1' 'b']
 ['2' 'b']
 ['3' 'b']]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [37]:
a = np.random.randint(0, 5, (8, 3))
print(a)
b = np.random.randint(0, 5, (2, 2))
print(b)
c = (a[..., np.newaxis, np.newaxis] == b)
rows = (c.sum(axis=(1, 2, 3)) >= b.shape[1]).nonzero()[0]
print(rows)

[[1 2 2]
 [4 2 3]
 [3 0 4]
 [0 1 4]
 [4 0 0]
 [3 3 2]
 [2 0 0]
 [4 1 0]]
[[2 3]
 [1 0]]
[0 1 2 3 4 5 6 7]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [38]:
a = np.random.randint(0, 9, (10, 3))
print(a)
unequal_rows = a[~np.all(a[:, 0][:, np.newaxis] == a[:, 1:], axis=1)]
print(unequal_rows)

[[3 3 7]
 [0 0 1]
 [7 0 5]
 [1 1 5]
 [0 1 2]
 [0 3 2]
 [7 6 7]
 [7 8 7]
 [8 2 3]
 [2 2 2]]
[[3 3 7]
 [0 0 1]
 [7 0 5]
 [1 1 5]
 [0 1 2]
 [0 3 2]
 [7 6 7]
 [7 8 7]
 [8 2 3]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [39]:
a = np.array([[1, 2, 3],
              [4, 5, 6],
              [1, 2, 3],
              [7, 8, 9],
              [4, 5, 6]])

a = np.unique(a, axis=0)
print(a)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [40]:
a = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
result = 1
for i in range(min(len(a), len(a[0]))):
    if a[i][i] != 0:
        result *= a[i][i]
print(result)

a = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
result = np.prod(np.diagonal(a)[np.diagonal(a) != 0])
print(result)

3
3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [41]:
def multi_py(x, y):
    return sorted(x) == sorted(y)


x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

result = multi_py(x, y)
print(result)

def multi_np(x, y):
    return np.array_equal(np.sort(x), np.sort(y))


x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

result = multi_np(x, y)
print(result)

True
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [42]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
zero = x == 0
print(x[1:][zero[:-1]].max())

def find_max_after_zero(arr):
    max_after_zero = None
    found_zero = False

    for num in arr:
        if found_zero:
            if max_after_zero is None or num > max_after_zero:
                max_after_zero = num

        if num == 0:
            found_zero = True
        else:
            found_zero = False

    return max_after_zero

x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
result = find_max_after_zero(x)

print(result)


5
5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [43]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
unique_values, counts = np.unique(x, return_counts=True)
print([unique_values, counts])


def count_digits_python(x):
    unique_values = []
    counts = []

    for num in x:
        if isinstance(num, int):
            for digit in str(num):
                if digit not in unique_values:
                    unique_values.append(digit)
                    counts.append(1)
                else:
                    index = unique_values.index(digit)
                    counts[index] += 1

    return tuple(unique_values), tuple(counts)


x = [2, 2, 2, 3, 3, 3, 5]
result = count_digits_python(x)
print(result)

[array([2, 3, 5]), array([3, 3, 1])]
(('2', '3', '5'), (3, 3, 1))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [44]:
from scipy.spatial.distance import cdist
import random
import time

def manual_euclidean_distances(X, Y):
    distances = np.sqrt(np.sum((X[:, np.newaxis] - Y) ** 2, axis=-1))
    return distances

X = np.random.rand(100, 3)
Y = np.random.rand(50, 3)

start_time_manual = time.time()
manual_distances = manual_euclidean_distances(X, Y)
end_time_manual = time.time()
print("Время manual_euclidean_distances:", end_time_manual - start_time_manual)

start_time_cdist = time.time()
cdist_distances = cdist(X, Y)
end_time_cdist = time.time()
print("Время cdist:", end_time_cdist - start_time_cdist)

print("Совпадают ли результаты:", np.allclose(manual_distances, cdist_distances))

import math
from scipy.spatial.distance import cdist
import time

def manual_euclidean_distances(X, Y):
    distances = []
    for x in X:
        row_distances = [math.sqrt(sum((xi - yi) ** 2 for xi, yi in zip(x, y))) for y in Y]
        distances.append(row_distances)
    return distances

X = [[random.random() for _ in range(3)] for _ in range(100)]
Y = [[random.random() for _ in range(3)] for _ in range(50)]

start_time_manual = time.time()
manual_distances = manual_euclidean_distances(X, Y)
end_time_manual = time.time()
print("Время manual_euclidean_distances:", end_time_manual - start_time_manual)

start_time_cdist = time.time()
cdist_distances = cdist(X, Y)
end_time_cdist = time.time()
print("Время cdist:", end_time_cdist - start_time_cdist)

print("Совпадают ли результаты:", np.allclose(manual_distances, cdist_distances))

Время manual_euclidean_distances: 0.0003921985626220703
Время cdist: 0.0002048015594482422
Совпадают ли результаты: True
Время manual_euclidean_distances: 0.003875732421875
Время cdist: 0.00019288063049316406
Совпадают ли результаты: True


* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [45]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [46]:
calorie_stats = np.loadtxt("data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [47]:
average_calories = np.average(calorie_stats, axis=0) - 60
print(f"Среднее количество калорий у конкурентов выше на {round(average_calories)} кал")

Среднее количество калорий у конкурентов выше на 47 кал


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [48]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [49]:
median_calories = np.median(calorie_stats_sorted, axis=0)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [50]:
percentiles = np.percentile(calorie_stats_sorted, np.arange(0, 101, 10))

nth_percentile = next(p for p in percentiles if p > 60)

print(f"Наименьший процентиль, превышающий 60 калорий: {nth_percentile}")

Наименьший процентиль, превышающий 60 калорий: 90.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [51]:
more_calories = np.mean(calorie_stats_sorted > 60) * 100
print(f"Процент хлопьев с более чем 60 калориями: {round(more_calories)}%")

Процент хлопьев с более чем 60 калориями: 96%


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [52]:
calorie_std = np.std(calorie_stats_sorted)

print(f"Стандартное отклонение количества калорий: {round(calorie_std)}")

Стандартное отклонение количества калорий: 19


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Если наша задача - использовать вычисления Numpy для анализа этих данных и доказать, что **СrunchieMunchies** - самый здоровый выбор для потребителей.
Используя наш анализ мы можем подчеркнуть преимущества CrunchieMunchies среди конкурентов.
1. В одной порции CrunchieMunchies содержится 60 калорий, в отличие от конкурентов, где среднее количество калорий в одной порции выше на 47 кал
2. У большинства конкурентов среднее количество калорий больше среднего, а именно около 100 и более
3. Процент хлопьев с более чем 60 калориями равен 96%, что выделяет нас на рынке